Schematics: `ng generate container ... --state reducers/index.ts` issue with .ts extension in state import

<!-- ❤️ ngrx? Please consider supporting our collective: 👉  [donate](https://opencollective.com/ngrx/donate) -->

## I'm submitting a...
<!-- Check one of the following options with "x" -->
<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  <!-- Please search GitHub for a similar issue or PR before submitting -->
[ ] Feature request
[ ] Documentation issue or request
</code></pre>

## What is the current behavior?
<!-- Describe the current behavior. -->
When running a command (almost-example from [docs](https://github.com/ngrx/platform/blob/master/docs/schematics/container.md#examples)):
```
ng generate container FooPage --state reducers/index.ts
```
the imported ?state within the generated `src/app/foo-page/foo-page.container.ts` is in the form
```
import * as fromStore from '../reducers/index.ts';
```
which causes a typescript compilation error
```
src/app/foo-page/foo-page.component.ts(3,28): error TS2691: An import path cannot end with a '.ts' extension. Consider importing '../reducers/index' instead.
```

Generating the container with a correct path (i.e. `--state reducers/index` or `--state reducers`) fails with
```
The Specified state path //src/app/reducers/index does not exist
```

The compilation error is easy to fix manually after generating; nevertheless it is a nuisance.

## Expected behavior:
<!-- Describe what the desired behavior would be. -->
The import path of ?state within a freshly generated container is in a form
```
import * as fromStore from '../reducers/index'; // no .ts extension
```
or in case of `index.ts` file even nicer
```
import * as fromStore from '../reducers'; // no index.ts
```

## Minimal reproduction of the problem with instructions:
<!--
For bug reports please provide the *STEPS TO REPRODUCE* and if possible a *MINIMAL DEMO* of the problem via
http://plnkr.co/edit/tpl:757r6L
-->
- generate a new angular/cli project
- setup ngrx/schematics (including initial state setup) as instructed in [schematics docs](https://github.com/ngrx/platform/tree/master/docs/schematics)
- `ng generate container FooPage --state reducers/index.ts`
- `ng serve`

## Version of affected browser(s),operating system(s), npm, node and ngrx:
Arch linux
npm 6.0.1
node 10.1.0
angular 6.0.0
ngrx 6.0.0-beta.2

--- a/modules/effects/schematics-core/utility/find-module.ts
+++ b/modules/effects/schematics-core/utility/find-module.ts
@@ -98,7 +98,7 @@ export function buildRelativePath(from: string, to: string): string {
 
   // Remove file names (preserving destination)
   fromParts.pop();
-  const toFileName = toParts.pop();
+  const toFileName = convertToTypeScriptFileName(toParts.pop());
 
   const relativePath = relative(
     normalize(fromParts.join('/')),
@@ -116,5 +116,16 @@ export function buildRelativePath(from: string, to: string): string {
     pathPrefix += '/';
   }
 
-  return pathPrefix + (relativePath ? relativePath + '/' : '') + toFileName;
+  return toFileName
+    ? pathPrefix + (relativePath ? relativePath + '/' : '') + toFileName
+    : pathPrefix + relativePath;
+}
+
+/**
+ * Strips the typescript extension and clears index filenames
+ * foo.ts -> foo
+ * index.ts -> empty
+ */
+function convertToTypeScriptFileName(filename: string | undefined) {
+  return filename ? filename.replace(/(\.ts)|(index\.ts)$/, '') : '';
 }

--- a/modules/entity/schematics-core/utility/find-module.ts
+++ b/modules/entity/schematics-core/utility/find-module.ts
@@ -98,7 +98,7 @@ export function buildRelativePath(from: string, to: string): string {
 
   // Remove file names (preserving destination)
   fromParts.pop();
-  const toFileName = toParts.pop();
+  const toFileName = convertToTypeScriptFileName(toParts.pop());
 
   const relativePath = relative(
     normalize(fromParts.join('/')),
@@ -116,5 +116,16 @@ export function buildRelativePath(from: string, to: string): string {
     pathPrefix += '/';
   }
 
-  return pathPrefix + (relativePath ? relativePath + '/' : '') + toFileName;
+  return toFileName
+    ? pathPrefix + (relativePath ? relativePath + '/' : '') + toFileName
+    : pathPrefix + relativePath;
+}
+
+/**
+ * Strips the typescript extension and clears index filenames
+ * foo.ts -> foo
+ * index.ts -> empty
+ */
+function convertToTypeScriptFileName(filename: string | undefined) {
+  return filename ? filename.replace(/(\.ts)|(index\.ts)$/, '') : '';
 }

--- a/modules/router-store/schematics-core/utility/find-module.ts
+++ b/modules/router-store/schematics-core/utility/find-module.ts
@@ -98,7 +98,7 @@ export function buildRelativePath(from: string, to: string): string {
 
   // Remove file names (preserving destination)
   fromParts.pop();
-  const toFileName = toParts.pop();
+  const toFileName = convertToTypeScriptFileName(toParts.pop());
 
   const relativePath = relative(
     normalize(fromParts.join('/')),
@@ -116,5 +116,16 @@ export function buildRelativePath(from: string, to: string): string {
     pathPrefix += '/';
   }
 
-  return pathPrefix + (relativePath ? relativePath + '/' : '') + toFileName;
+  return toFileName
+    ? pathPrefix + (relativePath ? relativePath + '/' : '') + toFileName
+    : pathPrefix + relativePath;
+}
+
+/**
+ * Strips the typescript extension and clears index filenames
+ * foo.ts -> foo
+ * index.ts -> empty
+ */
+function convertToTypeScriptFileName(filename: string | undefined) {
+  return filename ? filename.replace(/(\.ts)|(index\.ts)$/, '') : '';
 }

--- a/modules/schematics-core/utility/find-module.ts
+++ b/modules/schematics-core/utility/find-module.ts
@@ -98,7 +98,7 @@ export function buildRelativePath(from: string, to: string): string {
 
   // Remove file names (preserving destination)
   fromParts.pop();
-  const toFileName = toParts.pop();
+  const toFileName = convertToTypeScriptFileName(toParts.pop());
 
   const relativePath = relative(
     normalize(fromParts.join('/')),
@@ -116,5 +116,16 @@ export function buildRelativePath(from: string, to: string): string {
     pathPrefix += '/';
   }
 
-  return pathPrefix + (relativePath ? relativePath + '/' : '') + toFileName;
+  return toFileName
+    ? pathPrefix + (relativePath ? relativePath + '/' : '') + toFileName
+    : pathPrefix + relativePath;
+}
+
+/**
+ * Strips the typescript extension and clears index filenames
+ * foo.ts -> foo
+ * index.ts -> empty
+ */
+function convertToTypeScriptFileName(filename: string | undefined) {
+  return filename ? filename.replace(/(\.ts)|(index\.ts)$/, '') : '';
 }

--- a/modules/schematics/schematics-core/utility/find-module.ts
+++ b/modules/schematics/schematics-core/utility/find-module.ts
@@ -98,7 +98,7 @@ export function buildRelativePath(from: string, to: string): string {
 
   // Remove file names (preserving destination)
   fromParts.pop();
-  const toFileName = toParts.pop();
+  const toFileName = convertToTypeScriptFileName(toParts.pop());
 
   const relativePath = relative(
     normalize(fromParts.join('/')),
@@ -116,5 +116,16 @@ export function buildRelativePath(from: string, to: string): string {
     pathPrefix += '/';
   }
 
-  return pathPrefix + (relativePath ? relativePath + '/' : '') + toFileName;
+  return toFileName
+    ? pathPrefix + (relativePath ? relativePath + '/' : '') + toFileName
+    : pathPrefix + relativePath;
+}
+
+/**
+ * Strips the typescript extension and clears index filenames
+ * foo.ts -> foo
+ * index.ts -> empty
+ */
+function convertToTypeScriptFileName(filename: string | undefined) {
+  return filename ? filename.replace(/(\.ts)|(index\.ts)$/, '') : '';
 }

--- a/modules/schematics/src/container/index.spec.ts
+++ b/modules/schematics/src/container/index.spec.ts
@@ -57,6 +57,26 @@ describe('Container Schematic', () => {
     expect(content).toMatch(/import \* as fromStore from '..\/reducers';/);
   });
 
+  it('should remove .ts from the state path if provided', () => {
+    const options = { ...defaultOptions, state: 'reducers/foo.ts' };
+    appTree.create(`${projectPath}/src/app/reducers/foo.ts`, '');
+    const tree = schematicRunner.runSchematic('container', options, appTree);
+    const content = tree.readContent(
+      `${projectPath}/src/app/foo/foo.component.ts`
+    );
+    expect(content).toMatch(/import \* as fromStore from '..\/reducers\/foo';/);
+  });
+
+  it('should remove index.ts from the state path if provided', () => {
+    const options = { ...defaultOptions, state: 'reducers/index.ts' };
+    appTree.create(`${projectPath}/src/app/reducers/index.ts`, '');
+    const tree = schematicRunner.runSchematic('container', options, appTree);
+    const content = tree.readContent(
+      `${projectPath}/src/app/foo/foo.component.ts`
+    );
+    expect(content).toMatch(/import \* as fromStore from '..\/reducers';/);
+  });
+
   it('should import Store into the component', () => {
     const options = { ...defaultOptions, state: 'reducers' };
     appTree.create(`${projectPath}/src/app/reducers`, '');

--- a/modules/store-devtools/schematics-core/utility/find-module.ts
+++ b/modules/store-devtools/schematics-core/utility/find-module.ts
@@ -98,7 +98,7 @@ export function buildRelativePath(from: string, to: string): string {
 
   // Remove file names (preserving destination)
   fromParts.pop();
-  const toFileName = toParts.pop();
+  const toFileName = convertToTypeScriptFileName(toParts.pop());
 
   const relativePath = relative(
     normalize(fromParts.join('/')),
@@ -116,5 +116,16 @@ export function buildRelativePath(from: string, to: string): string {
     pathPrefix += '/';
   }
 
-  return pathPrefix + (relativePath ? relativePath + '/' : '') + toFileName;
+  return toFileName
+    ? pathPrefix + (relativePath ? relativePath + '/' : '') + toFileName
+    : pathPrefix + relativePath;
+}
+
+/**
+ * Strips the typescript extension and clears index filenames
+ * foo.ts -> foo
+ * index.ts -> empty
+ */
+function convertToTypeScriptFileName(filename: string | undefined) {
+  return filename ? filename.replace(/(\.ts)|(index\.ts)$/, '') : '';
 }

--- a/modules/store/schematics-core/utility/find-module.ts
+++ b/modules/store/schematics-core/utility/find-module.ts
@@ -98,7 +98,7 @@ export function buildRelativePath(from: string, to: string): string {
 
   // Remove file names (preserving destination)
   fromParts.pop();
-  const toFileName = toParts.pop();
+  const toFileName = convertToTypeScriptFileName(toParts.pop());
 
   const relativePath = relative(
     normalize(fromParts.join('/')),
@@ -116,5 +116,16 @@ export function buildRelativePath(from: string, to: string): string {
     pathPrefix += '/';
   }
 
-  return pathPrefix + (relativePath ? relativePath + '/' : '') + toFileName;
+  return toFileName
+    ? pathPrefix + (relativePath ? relativePath + '/' : '') + toFileName
+    : pathPrefix + relativePath;
+}
+
+/**
+ * Strips the typescript extension and clears index filenames
+ * foo.ts -> foo
+ * index.ts -> empty
+ */
+function convertToTypeScriptFileName(filename: string | undefined) {
+  return filename ? filename.replace(/(\.ts)|(index\.ts)$/, '') : '';
 }

